### PR TITLE
Add advanced location scanning

### PR DESCRIPTION
**Pull request by wilsonpage**

### CODE REVIEW

**Comment by wilsonpage** [comment]
@arcturus f?

File: src/main/java/org/mozilla/magnet/scanner/MagnetScanner.java [hunk 1]
```diff
@@ -31,11 +46,23 @@
  *
  * @author Francisco Jordano
  */
-public class MagnetScanner {
+public class MagnetScanner implements GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener, MagnetScannerListener {
```
Comment (arcturus) [inline src/main/java/org/mozilla/magnet/scanner/MagnetScanner.java]: nit: GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener -> OnConnectionFailedListener

Comment (arcturus) [inline src/main/java/org/mozilla/magnet/scanner/MagnetScanner.java]: Is a bit weird to me this class implements MagnetScannerListener, when it's aggregating different scanners.

File: src/main/java/org/mozilla/magnet/scanner/ReceiverBackgroundScanner.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        Log.d(TAG, "on receive: " + intent.getAction());
+        Intent serviceIntent = new Intent(context, ServiceBackgroundScanner.class);
+
+        Bundle bundle = intent.getExtras();
+        long timeStarted = bundle.getLong("started");
+        long elapsed = System.currentTimeMillis() - timeStarted;
+
+        // the background requestLocationUpdates fires as soon as
+        // it's scheduled; we ignore any broadcasts that happen
+        // 'too soon' after the app haes gone to the background
```
Comment (arcturus) [inline src/main/java/org/mozilla/magnet/scanner/ReceiverBackgroundScanner.java:33]: nit: 'has'

File: src/main/java/org/mozilla/magnet/scanner/ServiceBackgroundScanner.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    /**
+     * Called once, when the service is terminated.
+     */
+    @Override
+    public void onDestroy() {
+        Log.d(TAG, "on destroy");
+        stop();
+    }
+
+    public void addStopListener() {
```
Comment (arcturus) [inline src/main/java/org/mozilla/magnet/scanner/ServiceBackgroundScanner.java]: nit: do we need this 'public'?

File: src/main/java/org/mozilla/magnet/scanner/ServiceBackgroundScanner.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        LocalBroadcastManager
+                .getInstance(this)
+                .registerReceiver(mReceiver, new IntentFilter(MagnetScanner.ACTION_STOP_SCAN));
+    }
+
+    public void removeStopListener() {
```
Comment (arcturus) [inline src/main/java/org/mozilla/magnet/scanner/ServiceBackgroundScanner.java]: nit: same here, the visibility doesn't need to be public right?

File: src/main/java/org/mozilla/magnet/scanner/ServiceBackgroundScanner.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    private BroadcastReceiver mReceiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            stop();
+        }
+    };
+
+    public void broadcastFoundItems(ArrayList<MagnetScannerItem> items) {
```
Comment (arcturus) [inline src/main/java/org/mozilla/magnet/scanner/ServiceBackgroundScanner.java]: nit: visibility of this method

File: src/main/java/org/mozilla/magnet/scanner/ble/ScannerBLE.java [hunk 6]
```diff
@@ -19,111 +22,111 @@
...
-            this.maxSize = maxSize;
-        }
+    private void scheduleExpireCheck() {
+        mHandler.postDelayed(expireCheck, EXPIRE_CHECK_INTERVAL_MS);
+    }
 
-        @Override
-        protected boolean removeEldestEntry(Map.Entry<K, V> eldest) {
-            return size() > maxSize;
+    private Runnable expireCheck = new Runnable() {
```
Comment (arcturus) [inline src/main/java/org/mozilla/magnet/scanner/ble/ScannerBLE.java:108]: This is pretty cool!

**Review by arcturus — COMMENTED** [review]
Looking pretty good, added some nits, still need to try on the device.

**Review by arcturus — APPROVED** [review]
Looking good to me, shiiiip it!

**Pull request merged**